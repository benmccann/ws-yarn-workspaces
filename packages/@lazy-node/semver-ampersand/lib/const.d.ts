export declare const reAmpersandAndSpaces: RegExp;
export declare const reInvalidCharacter: RegExp;
