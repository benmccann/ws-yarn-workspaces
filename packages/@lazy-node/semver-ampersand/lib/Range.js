"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Range = void 0;
const range_1 = __importDefault(require("semver/classes/range"));
const handleVersionRange_1 = require("./handleVersionRange");
class Range extends range_1.default {
    constructor(rawRange, optionsOrLoose) {
        const range = handleVersionRange_1.handleVersionRange(rawRange);
        super(range, optionsOrLoose);
        if (typeof rawRange === 'string' && range !== rawRange) {
            this.rawRange = rawRange;
        }
    }
}
exports.Range = Range;
//# sourceMappingURL=Range.js.map