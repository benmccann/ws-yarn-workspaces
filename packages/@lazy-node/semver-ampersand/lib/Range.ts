import semverRange from 'semver/classes/range';
import { Options } from 'semver';
import { handleVersionRange } from './handleVersionRange';

export class SemverRange extends semverRange
{
	rawSource?: string | Range;

	constructor(rawSource: string | SemverRange | semverRange, optionsOrLoose?: boolean | Options) {
		const range = handleVersionRange(rawSource);

		super(range, optionsOrLoose);

		if (typeof rawSource === 'string' && range !== rawSource)
		{
			this.rawSource = rawSource;
		}
	}

}

export { SemverRange as Range }

export default SemverRange
