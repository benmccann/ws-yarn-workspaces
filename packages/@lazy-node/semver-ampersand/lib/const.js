"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EnumSemverVersion = exports.separatorDoubleVerticalBar = exports.reDoubleVerticalBar = exports.reInvalidCharacterOrAmpersand = exports.reInvalidCharacter = exports.reAmpersandAndSpaces = exports.reSpaces = void 0;
exports.reSpaces = /\s+/g;
exports.reAmpersandAndSpaces = /[&\s]+/g;
exports.reInvalidCharacter = /[^\s\w.*\-><~!^|=+]/;
exports.reInvalidCharacterOrAmpersand = /[^\s\w.*\-><~!^|=+&]/;
exports.reDoubleVerticalBar = /\s*\|\|\s*/g;
exports.separatorDoubleVerticalBar = ' || ';
var EnumSemverVersion;
(function (EnumSemverVersion) {
    /**
     * nothing is allowed
     */
    EnumSemverVersion["NULL"] = "<0.0.0-0";
    /**
     * nothing is forbidden
     */
    EnumSemverVersion["STAR"] = "*";
    EnumSemverVersion["ANY"] = "";
})(EnumSemverVersion = exports.EnumSemverVersion || (exports.EnumSemverVersion = {}));
//# sourceMappingURL=const.js.map