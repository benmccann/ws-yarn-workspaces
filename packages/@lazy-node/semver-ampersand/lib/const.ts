
export const reSpaces = /\s+/g;

export const reAmpersandAndSpaces = /[&\s]+/g;

export const reInvalidCharacter = /[^\s\w.*\-><~!^|=+]/;

export const reInvalidCharacterOrAmpersand = /[^\s\w.*\-><~!^|=+&]/;

export const reDoubleVerticalBar = /\s*\|\|\s*/g;

export const separatorDoubleVerticalBar = ' || ' as const;
