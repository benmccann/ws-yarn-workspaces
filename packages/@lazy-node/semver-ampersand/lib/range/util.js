"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.hyphenReplace = hyphenReplace;
exports.isX = isX;
exports.replaceGTE0 = replaceGTE0;
exports.parseComparator = parseComparator;
exports.replaceTildes = replaceTildes;
exports.replaceTilde = replaceTilde;
exports.replaceCarets = replaceCarets;
exports.replaceCaret = replaceCaret;
exports.replaceXRanges = replaceXRanges;
exports.replaceXRange = replaceXRange;
exports.replaceStars = replaceStars;
const re_1 = require("semver/internal/re");
/**
 * This function is passed to string.replace(re[t.HYPHENRANGE])
 * M, m, patch, prerelease, build
 * 1.2 - 3.4.5 => >=1.2.0 <=3.4.5
 * 1.2.3 - 3.4 => >=1.2.0 <3.5.0-0 Any 3.4.x will do
 * 1.2 - 3.4 => >=1.2.0 <3.5.0-0
 */
function hyphenReplace(incPr) {
    return ($0, from, fM, fm, fp, fpr, fb, to, tM, tm, tp, tpr, tb) => {
        if (isX(fM)) {
            from = '';
        }
        else if (isX(fm)) {
            from = `>=${fM}.0.0${incPr ? '-0' : ''}`;
        }
        else if (isX(fp)) {
            from = `>=${fM}.${fm}.0${incPr ? '-0' : ''}`;
        }
        else if (fpr) {
            from = `>=${from}`;
        }
        else {
            from = `>=${from}${incPr ? '-0' : ''}`;
        }
        if (isX(tM)) {
            to = '';
        }
        else if (isX(tm)) {
            to = `<${+tM + 1}.0.0-0`;
        }
        else if (isX(tp)) {
            to = `<${tM}.${+tm + 1}.0-0`;
        }
        else if (tpr) {
            to = `<=${tM}.${tm}.${tp}-${tpr}`;
        }
        else if (incPr) {
            to = `<${tM}.${tm}.${+tp + 1}-0`;
        }
        else {
            to = `<=${to}`;
        }
        return (`${from} ${to}`).trim();
    };
}
function isX(id) {
    return !id || id.toLowerCase() === 'x' || id === '*';
}
function replaceGTE0(comp, options) {
    return comp.trim()
        .replace(re_1.re[(options === null || options === void 0 ? void 0 : options.includePrerelease) ? re_1.t.GTE0PRE : re_1.t.GTE0], '');
}
/**
 * comprised of xranges, tildes, stars, and gtlt's at this point.
 * already replaced the hyphen ranges
 * turn into a set of JUST comparators.
 */
function parseComparator(comp, options) {
    comp = replaceCarets(comp, options);
    comp = replaceTildes(comp, options);
    comp = replaceXRanges(comp, options);
    comp = replaceStars(comp, options);
    return comp;
}
/**
 * ~, ~> --> * (any, kinda silly)
 * ~2, ~2.x, ~2.x.x, ~>2, ~>2.x ~>2.x.x --> >=2.0.0 <3.0.0-0
 * ~2.0, ~2.0.x, ~>2.0, ~>2.0.x --> >=2.0.0 <2.1.0-0
 * ~1.2, ~1.2.x, ~>1.2, ~>1.2.x --> >=1.2.0 <1.3.0-0
 */
function replaceTildes(comp, options) {
    return comp.trim().split(/\s+/).map((comp) => {
        return replaceTilde(comp, options);
    }).join(' ');
}
function replaceTilde(comp, options) {
    const r = options.loose ? re_1.re[re_1.t.TILDELOOSE] : re_1.re[re_1.t.TILDE];
    return comp.replace(r, (_, M, m, p, pr) => {
        let ret;
        if (isX(M)) {
            ret = '';
        }
        else if (isX(m)) {
            ret = `>=${M}.0.0 <${+M + 1}.0.0-0`;
        }
        else if (isX(p)) {
            // ~1.2 == >=1.2.0 <1.3.0-0
            ret = `>=${M}.${m}.0 <${M}.${+m + 1}.0-0`;
        }
        else if (pr) {
            ret = `>=${M}.${m}.${p}-${pr} <${M}.${+m + 1}.0-0`;
        }
        else {
            // ~1.2.3 == >=1.2.3 <1.3.0-0
            ret = `>=${M}.${m}.${p} <${M}.${+m + 1}.0-0`;
        }
        return ret;
    });
}
/**
 * ^ --> * (any, kinda silly)
 * ^2, ^2.x, ^2.x.x --> >=2.0.0 <3.0.0-0
 * ^2.0, ^2.0.x --> >=2.0.0 <3.0.0-0
 * ^1.2, ^1.2.x --> >=1.2.0 <2.0.0-0
 * ^1.2.3 --> >=1.2.3 <2.0.0-0
 * ^1.2.0 --> >=1.2.0 <2.0.0-0
 */
function replaceCarets(comp, options) {
    return comp.trim().split(/\s+/).map((comp) => {
        return replaceCaret(comp, options);
    }).join(' ');
}
function replaceCaret(comp, options) {
    const r = options.loose ? re_1.re[re_1.t.CARETLOOSE] : re_1.re[re_1.t.CARET];
    const z = options.includePrerelease ? '-0' : '';
    return comp.replace(r, (_, M, m, p, pr) => {
        let ret;
        if (isX(M)) {
            ret = '';
        }
        else if (isX(m)) {
            ret = `>=${M}.0.0${z} <${+M + 1}.0.0-0`;
        }
        else if (isX(p)) {
            if (M === '0') {
                ret = `>=${M}.${m}.0${z} <${M}.${+m + 1}.0-0`;
            }
            else {
                ret = `>=${M}.${m}.0${z} <${+M + 1}.0.0-0`;
            }
        }
        else if (pr) {
            if (M === '0') {
                if (m === '0') {
                    ret = `>=${M}.${m}.${p}-${pr} <${M}.${m}.${+p + 1}-0`;
                }
                else {
                    ret = `>=${M}.${m}.${p}-${pr} <${M}.${+m + 1}.0-0`;
                }
            }
            else {
                ret = `>=${M}.${m}.${p}-${pr} <${+M + 1}.0.0-0`;
            }
        }
        else {
            if (M === '0') {
                if (m === '0') {
                    ret = `>=${M}.${m}.${p}${z} <${M}.${m}.${+p + 1}-0`;
                }
                else {
                    ret = `>=${M}.${m}.${p}${z} <${M}.${+m + 1}.0-0`;
                }
            }
            else {
                ret = `>=${M}.${m}.${p} <${+M + 1}.0.0-0`;
            }
        }
        return ret;
    });
}
function replaceXRanges(comp, options) {
    return comp.split(/\s+/).map((comp) => {
        return replaceXRange(comp, options);
    }).join(' ');
}
function replaceXRange(comp, options) {
    comp = comp.trim();
    const r = options.loose ? re_1.re[re_1.t.XRANGELOOSE] : re_1.re[re_1.t.XRANGE];
    return comp.replace(r, (ret, gtlt, M, m, p, pr) => {
        const xM = isX(M);
        const xm = xM || isX(m);
        const xp = xm || isX(p);
        const anyX = xp;
        if (gtlt === '=' && anyX) {
            gtlt = '';
        }
        // if we're including prereleases in the match, then we need
        // to fix this to -0, the lowest possible prerelease value
        pr = options.includePrerelease ? '-0' : '';
        if (xM) {
            if (gtlt === '>' || gtlt === '<') {
                // nothing is allowed
                ret = "<0.0.0-0" /* EnumSemverVersion.NULL */;
            }
            else {
                // nothing is forbidden
                ret = "*" /* EnumSemverVersion.STAR */;
            }
        }
        else if (gtlt && anyX) {
            // we know patch is an x, because we have any x at all.
            // replace X with 0
            if (xm) {
                m = 0;
            }
            p = 0;
            if (gtlt === '>') {
                // >1 => >=2.0.0
                // >1.2 => >=1.3.0
                gtlt = '>=';
                if (xm) {
                    M = +M + 1;
                    m = 0;
                    p = 0;
                }
                else {
                    m = +m + 1;
                    p = 0;
                }
            }
            else if (gtlt === '<=') {
                // <=0.7.x is actually <0.8.0, since any 0.7.x should
                // pass.  Similarly, <=7.x is actually <8.0.0, etc.
                gtlt = '<';
                if (xm) {
                    M = +M + 1;
                }
                else {
                    m = +m + 1;
                }
            }
            if (gtlt === '<') {
                pr = '-0';
            }
            ret = `${gtlt + M}.${m}.${p}${pr}`;
        }
        else if (xm) {
            ret = `>=${M}.0.0${pr} <${+M + 1}.0.0-0`;
        }
        else if (xp) {
            ret = `>=${M}.${m}.0${pr} <${M}.${+m + 1}.0-0`;
        }
        return ret;
    });
}
/**
 * and '' means "any version", just remove the *s entirely.
 */
function replaceStars(comp, options) {
    // Looseness is ignored here.  star is always as loose as it gets!
    return comp.trim().replace(re_1.re[re_1.t.STAR], '');
}
//# sourceMappingURL=util.js.map