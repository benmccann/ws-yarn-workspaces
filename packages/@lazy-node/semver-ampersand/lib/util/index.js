"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isPlainObject = exports.hasAmpersandAndSpaces = void 0;
const const_1 = require("../const");
function hasAmpersandAndSpaces(input) {
    return const_1.reAmpersandAndSpaces.test(input);
}
exports.hasAmpersandAndSpaces = hasAmpersandAndSpaces;
function isPlainObject(value) {
    if (Object.prototype.toString.call(value) !== '[object Object]') {
        return false;
    }
    const prototype = Object.getPrototypeOf(value);
    return prototype === null || prototype === Object.prototype;
}
exports.isPlainObject = isPlainObject;
//# sourceMappingURL=index.js.map