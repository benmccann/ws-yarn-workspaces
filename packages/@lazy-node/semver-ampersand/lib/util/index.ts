import { reAmpersandAndSpaces } from '../const';

export function hasAmpersandAndSpaces(input: string)
{
	return reAmpersandAndSpaces.test(input)
}

export function isPlainObject(value): value is object
{
	if (Object.prototype.toString.call(value) !== '[object Object]')
	{
		return false;
	}

	const prototype = Object.getPrototypeOf(value);
	return prototype === null || prototype === Object.prototype;
}
