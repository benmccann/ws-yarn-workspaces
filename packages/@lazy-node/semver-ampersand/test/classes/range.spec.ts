import { Comparator } from 'semver';
import { SemverRange, Range } from '../../lib/Range';

import rangeIntersection from '../fixtures/range-intersection'
import rangeInclude from '../fixtures/range-include'
import rangeExclude from '../fixtures/range-exclude'
import rangeParse from '../fixtures/range-parse'
import { reDoubleVerticalBar, separatorDoubleVerticalBar } from '../../lib/const';

import { inspect } from 'util';

//jest.setTimeout(5000 * 10);

describe('range tests', () =>
{
	rangeInclude.forEach(([range, ver, options]) =>
	{
		test(`${range} ${inspect(options)}`, () => {
			const r = new Range(range, options)
			expect(r.test(ver)).toBeTruthy()
		})
	})
})

describe('range parsing', () =>
{
	//expect.assertions(rangeParse.length)
	rangeParse.forEach(([range, expected, options]) => test(`${range} ${inspect(options)} ${expected = _replaceDoubleVerticalBar(expected)}`, () =>
	{
		if (expected == null)
		{
			expect(() => new Range(range)).toThrowErrorMatchingSnapshot()
		}
		else
		{
			expect(new Range(range, options).range || '*').toStrictEqual(expected)
			expect(new Range(range, options).range).toStrictEqual(new Range(expected).range)
		}
	}))
})

test('throw for empty comparator set, even in loose mode', () =>
{
	expect(() => new Range('sadf||asdf', { loose: true })).toThrowError(TypeError('Invalid SemVer Range: sadf||asdf'))
})

test('convert comparator to range', () =>
{
	const c = new Comparator('>=1.2.3')
	const r = new Range(c)
	expect(r.raw).toStrictEqual(c.value)
})

test('range as argument to range ctor', () =>
{
	const loose = new Range('1.2.3', { loose: true })

	expect(loose).toMatchSnapshot();

	expect(new Range(loose, { loose: true })).toStrictEqual(loose)
	expect(new Range(loose, true)).toStrictEqual(loose)

	expect(new Range(loose)).toStrictEqual(loose)
	expect(new Range(loose, { loose: false })).not.toStrictEqual(loose)

	const incPre = new Range('1.2.3', { includePrerelease: true })

	expect(incPre).toMatchSnapshot();

	expect(new Range(incPre, { includePrerelease: true })).toStrictEqual(incPre)
	expect(new Range(incPre)).toStrictEqual(incPre)
})

describe('negative range tests', () =>
{
	rangeExclude.forEach(([range, ver, options]) =>
	{
		test(`${range} ${inspect(options)}`, () => {
			const r = new Range(range, options)
			expect(r.test(ver)).toBeFalsy()
		})
	})
})

test('strict vs loose ranges', () =>
{
	[
		['>=01.02.03', '>=1.2.3'],
		['~1.02.03beta', '>=1.2.3-beta <1.3.0-0'],
	].forEach(([loose, comps]) =>
	{
		expect(() => new Range(loose)).toThrow()
		expect(new Range(loose, true).range).toStrictEqual(comps)
	})
})

test('tostrings', () =>
{
	expect(new Range('>= v1.2.3').toString()).toStrictEqual('>=1.2.3')
})

describe('ranges intersect', () =>
{
	rangeIntersection.forEach(([r0, r1, expected]) =>
	{
		test(`${r0} <~> ${r1}`, () =>
		{
			const range0 = new Range(r0)
			const range1 = new Range(r1)

			expect(range0.intersects(range1)).toStrictEqual(expected)
			expect(range1.intersects(range0)).toStrictEqual(expected)
		})
	})
})

test('missing range parameter in range intersect', () =>
{
	expect(() =>
	{
		// @ts-expect-error
		new Range('1.0.0').intersects()
	}).toThrowError(new TypeError('a Range is required'))
})

function _replaceDoubleVerticalBar(expected: string)
{
	if (expected != null)
	{
		return expected.replace(reDoubleVerticalBar, separatorDoubleVerticalBar)
	}

	return expected
}
