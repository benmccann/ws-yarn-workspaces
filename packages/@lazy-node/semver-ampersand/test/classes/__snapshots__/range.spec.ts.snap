// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`range as argument to range ctor 1`] = `
SemverRange {
  "includePrerelease": false,
  "loose": true,
  "options": Object {
    "includePrerelease": false,
    "loose": true,
  },
  "range": "1.2.3",
  "raw": "1.2.3",
  "set": Array [
    Array [
      Comparator {
        "loose": true,
        "operator": "",
        "options": Object {
          "loose": true,
        },
        "semver": SemVer {
          "build": Array [],
          "includePrerelease": false,
          "loose": true,
          "major": 1,
          "minor": 2,
          "options": Object {
            "loose": true,
          },
          "patch": 3,
          "prerelease": Array [],
          "raw": "1.2.3",
          "version": "1.2.3",
        },
        "value": "1.2.3",
      },
    ],
  ],
}
`;

exports[`range as argument to range ctor 2`] = `
SemverRange {
  "includePrerelease": true,
  "loose": false,
  "options": Object {
    "includePrerelease": true,
    "loose": false,
  },
  "range": "1.2.3",
  "raw": "1.2.3",
  "set": Array [
    Array [
      Comparator {
        "loose": false,
        "operator": "",
        "options": Object {
          "includePrerelease": true,
        },
        "semver": SemVer {
          "build": Array [],
          "includePrerelease": false,
          "loose": false,
          "major": 1,
          "minor": 2,
          "options": Object {},
          "patch": 3,
          "prerelease": Array [],
          "raw": "1.2.3",
          "version": "1.2.3",
        },
        "value": "1.2.3",
      },
    ],
  ],
}
`;

exports[`range parsing >01.02.03 undefined null 1`] = `"Invalid comparator: >01.02.03"`;

exports[`range parsing ~1.2.3beta undefined null 1`] = `"Invalid comparator: ~1.2.3beta"`;
